### PR TITLE
add `DisplayLinkUpdateActiveMonitor` to MTLGraphicsDevice for updating `CVDisplayLink` current display

### PR DESCRIPTION
**Pull request by ILW8**
This allows a user of the renderer to update the CVDisplayLink's current display by passing the active display's origin coordinates and size. 

This is needed to correctly set CVDisplayLink's target refresh rate when:
- the window is moved across monitors
- CVDisplayLinkCreateWithActiveCGDisplays (at [Veldrid.MetalBindings/CVDisplayLink.cs#L21](https://github.com/ILW8/veldrid/blob/542b3f255923610b8fb4e114f3f832ee3cbe0cbd/src/Veldrid.MetalBindings/CVDisplayLink.cs#LL21C13-L21C76)) doesn't return a CVDisplayLink set to the correct refresh rate; I've noticed this is especially prevalent when the high refresh rate monitor is mirrored to another lower refresh rate monitor, or in the case where the main monitor is 60Hz but a game using Veldrid is being run on a different high refresh rate monitor.

I'm using origin and size because it's the only relatively straightforward way of finding the correct [CGDirectDisplayID](https://developer.apple.com/documentation/coregraphics/cgdirectdisplayid) for calling [CVDisplayLinkSetCurrentCGDisplay](https://developer.apple.com/documentation/corevideo/1456768-cvdisplaylinksetcurrentcgdisplay) with (that I could find). 

I feel that the way things are laid out (having such a deep call stack until actually calling CVDisplayLinkSetCurrentCGDisplay) is not great, but I'm not sure how else this should be laid out. I wouldn't mind if this code is completely thrown out in favour of something else, but at least this code should give an idea of what needs to be done.

### CODE REVIEW

File: src/Veldrid/GraphicsDevice.cs [hunk 1]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
+        }
+
+        /// <summary>
+        /// Calls CVDisplayLinkGetActualOutputVideoRefreshPeriod
+        /// </summary>
+        /// <returns>The actual refresh period of the currently selected CVDisplayLink display in seconds</returns>
+        public virtual double DisplayLinkGetActualOutputVideoRefreshPeriod()
+        {
+            return -1.0f;
+        }
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:990]: When is this going to be used?

File: src/Veldrid/GraphicsDevice.cs [hunk 2]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
         }
+
+        /// <summary>
+        /// Updates CVDisplayLink active display using the display's origin and size.
+        /// </summary>
+        /// <param name="x"></param> x coordinate of the display's origin
+        /// <param name="y"></param> y coordinate of the display's origin
+        /// <param name="w"></param> display's width
+        /// <param name="h"></param> display's height
+        public virtual void DisplayLinkUpdateActiveMonitor(int x, int y, int w, int h)
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:978]: I'd remove the `DisplayLink` prefix, maybe it'll be useful for other graphics devices.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: src/Veldrid/GraphicsDevice.cs (hunk 1 above)
Comment (ILW8) [reply src/Veldrid/GraphicsDevice.cs:990]: Planning to use this in a visual test. A monitor's refresh rate can be accessed as osu.Framework.Platform.Display.DisplayModes.FirstOrDefault(), then we can check that CVDisplayLink's refresh period matches (closely enough) to the expected monitor's refresh rate.

File: src/Veldrid/GraphicsDevice.cs
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:1]: Normally I'd say to revert the rest of the unnecessary changes to this file, but I also deal with them locally every time, so I think it's fine.

**Comment by smoogipoo** [comment]
Looks good to me, is this ready for review?

File: src/Veldrid/GraphicsDevice.cs [hunk 3]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
         }
+
+        /// <summary>
+        /// Updates CVDisplayLink active display using the display's origin and size.
+        /// </summary>
+        /// <param name="x"></param> x coordinate of the display's origin
+        /// <param name="y"></param> y coordinate of the display's origin
+        /// <param name="w"></param> display's width
+        /// <param name="h"></param> display's height
+        public virtual void UpdateActiveMonitor(int x, int y, int w, int h)
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:978]: Would rename to `UpdateActiveDisplay`.

File: src/Veldrid/GraphicsDevice.cs [hunk 4]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
 
             return info;
         }
+
+        /// <summary>
+        /// Updates CVDisplayLink active display using the display's origin and size.
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:972]:
```suggestion
        /// Updates the active display using the display's origin and size.
```

File: src/Veldrid/GraphicsDevice.cs [hunk 5]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
+            return;
+        }
+
+        /// <summary>
+        /// Calls CVDisplayLinkGetActualOutputVideoRefreshPeriod
+        /// </summary>
+        /// <returns>The actual refresh period of the currently selected CVDisplayLink display in seconds</returns>
+        public virtual double DisplayLinkGetActualOutputVideoRefreshPeriod()
+        {
+            return -1.0f;
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:989]: Why -1? Even `CVDisplayLinkGetActualOutputVideoRefreshPeriod` returns 0 if not running.

File: src/Veldrid/GraphicsDevice.cs [hunk 6]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
+        /// <param name="y"></param> y coordinate of the display's origin
+        /// <param name="w"></param> display's width
+        /// <param name="h"></param> display's height
+        public virtual void UpdateActiveMonitor(int x, int y, int w, int h)
+        {
+            return;
+        }
+
+        /// <summary>
+        /// Calls CVDisplayLinkGetActualOutputVideoRefreshPeriod
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:984]: Remove reference to `CVDisplayLinkGetActualOutputVideoRefreshPeriod` since this is on the base `GraphicsDevice`.

File: src/Veldrid/GraphicsDevice.cs [hunk 7]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
+        /// </summary>
+        /// <returns>The actual refresh period of the currently selected CVDisplayLink display in seconds</returns>
+        public virtual double DisplayLinkGetActualOutputVideoRefreshPeriod()
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:987]: How about `GetActualRefreshPeriod()`?

I'm also contemplating whether this should even exist in the first place. It's likely not going to be implementable on anything else but MTL, and is only to be used for logging. I would rather it not exist. @peppy any thoughts?

File: src/Veldrid/GraphicsDevice.cs [hunk 8]
```diff
@@ -967,6 +967,27 @@ public BackendInfoMetal GetMetalInfo()
...
+        /// <param name="h"></param> display's height
+        public virtual void UpdateActiveMonitor(int x, int y, int w, int h)
+        {
+            return;
+        }
+
+        /// <summary>
+        /// Calls CVDisplayLinkGetActualOutputVideoRefreshPeriod
+        /// </summary>
+        /// <returns>The actual refresh period of the currently selected CVDisplayLink display in seconds</returns>
```
Comment (smoogipoo) [inline src/Veldrid/GraphicsDevice.cs:986]: Also remove `CVDisplayLink` here.

Is it seconds-per-refresh, or refreshes-per-second? Can we make it exactly `milliseconds per refresh`?

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Just some nits

File: src/Veldrid/GraphicsDevice.cs (hunk 5 above)
Comment (ILW8) [reply src/Veldrid/GraphicsDevice.cs:989]: The idea was that if `DisplayLinkGetActualOutputVideoRefreshPeriod` is called when not using MTL (which it shouldn't), a return value of -1 signifies some kind of error/something not applicable whereas 0 with DisplayLink means the display is not outputting. 
Am I overthinking it?

**Comment by peppy** [comment]
I've tested this code and it seems to hang for me:

JetBrains Rider 2023-06-15 at 10 42 47

@ILW8 did you manage to make this work with the framework diff you provided?

**Comment by ILW8** [comment]
Adding this line also works, but I'm not sure that's ideal (I don't really know what this does).
```diff
index 478a502..56de90b 100644
--- a/src/Veldrid/MTL/MTLGraphicsDevice.cs
+++ b/src/Veldrid/MTL/MTLGraphicsDevice.cs
@@ -62,6 +62,7 @@ namespace Veldrid.MTL
         {
             if (_displayLink != null)
             {
+                _frameEndedEvent.Set();
                 _displayLink.UpdateActiveMonitor(x, y, w, h);
             }
         }

```

File: src/Veldrid/MTL/MTLGraphicsDevice.cs [hunk 9]
```diff
@@ -58,6 +58,25 @@ public MTLGraphicsDevice(GraphicsDeviceOptions options, SwapchainDescription? sw
         {
         }
 
+        public override void UpdateActiveDisplay(int x, int y, int w, int h)
+        {
+            if (_displayLink != null)
+            {
+                _frameEndedEvent.Set();
```
Comment (frenzibyte) [inline src/Veldrid/MTL/MTLGraphicsDevice.cs:65]: This is definitely not the way to go about this. This could *potentially* cause permanent desynchronisation between callback and V-sync, at least according to my knowledge on how `CADisplayLink` works (not sure how that goes for `CVDisplayLink`). And is generally a hacky bandaid at best.

This suggests that the framework-side code isn't calling `UpdateActiveDisplay` at a good time, it must be called in the `DrawThread` in a safe place where the display link isn't blocked.

Here's an `osu-framework` diff that does this a bit smarter:
```diff
diff --git a/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs b/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
index cc7d06722..6eedb6bf3 100644
--- a/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
+++ b/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
@@ -199,8 +199,11 @@ protected override void Initialise(IGraphicsSurface graphicsSurface)
             BufferUpdateCommands = Factory.CreateCommandList();
 
             pipeline.Outputs = Device.SwapchainFramebuffer.OutputDescription;
+
+            graphicsSurface.ActiveDisplayChanged += d => pendingActiveDisplay = d;
         }
 
+        private Display? pendingActiveDisplay;
         private Vector2 currentSize;
 
         protected internal override void BeginFrame(Vector2 windowSize)
@@ -232,7 +235,17 @@ protected internal override void FinishFrame()
             Device.SubmitCommands(Commands);
         }
 
-        protected internal override void SwapBuffers() => Device.SwapBuffers();
+        protected internal override void SwapBuffers()
+        {
+            Device.SwapBuffers();
+
+            if (pendingActiveDisplay != null)
+            {
+                Device.UpdateActiveDisplay(pendingActiveDisplay.Bounds.X, pendingActiveDisplay.Bounds.Y, pendingActiveDisplay.Bounds.Width, pendingActiveDisplay.Bounds.Height);
+                pendingActiveDisplay = null;
+            }
+        }
+
         protected internal override void WaitUntilIdle() => Device.WaitForIdle();
         protected internal override void WaitUntilNextFrameReady() => Device.WaitForNextFrameReady();
 
diff --git a/osu.Framework/Platform/IGraphicsSurface.cs b/osu.Framework/Platform/IGraphicsSurface.cs
index cc4c31562..ab29ec095 100644
--- a/osu.Framework/Platform/IGraphicsSurface.cs
+++ b/osu.Framework/Platform/IGraphicsSurface.cs
@@ -11,6 +11,8 @@ namespace osu.Framework.Platform
     /// </summary>
     public interface IGraphicsSurface
     {
+        event Action<Display>? ActiveDisplayChanged;
+
         /// <summary>
         /// A pointer representing a handle to this window, provided by the operating system.
         /// </summary>
diff --git a/osu.Framework/Platform/OsuTKGraphicsSurface.cs b/osu.Framework/Platform/OsuTKGraphicsSurface.cs
index 11367fa4e..77999fc08 100644
--- a/osu.Framework/Platform/OsuTKGraphicsSurface.cs
+++ b/osu.Framework/Platform/OsuTKGraphicsSurface.cs
@@ -17,6 +17,7 @@ public class OsuTKGraphicsSurface : IGraphicsSurface, IOpenGLGraphicsSurface
 
         public GraphicsSurfaceType Type => GraphicsSurfaceType.OpenGL;
 
+        public event Action<Display> ActiveDisplayChanged;
         public IntPtr WindowHandle => window.WindowInfo.Handle;
         public IntPtr DisplayHandle => throw new NotSupportedException($@"{nameof(DisplayHandle)} is not supported.");
 
diff --git a/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs b/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs
index 0c13b56ef..e66b410c1 100644
--- a/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs
+++ b/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs
@@ -19,6 +19,8 @@ public class SDL2GraphicsSurface : IGraphicsSurface, IOpenGLGraphicsSurface, IMe
 
         private IntPtr context;
 
+        public event Action<Display>? ActiveDisplayChanged;
+
         public IntPtr WindowHandle => window.WindowHandle;
         public IntPtr DisplayHandle => window.DisplayHandle;
 
@@ -84,6 +86,11 @@ public Size GetDrawableSize()
             return new Size(width, height);
         }
 
+        public void UpdateActiveDisplay(Display display)
+        {
+            ActiveDisplayChanged?.Invoke(display);
+        }
+
         #region OpenGL-specific implementation
 
         private void initialiseOpenGL()
diff --git a/osu.Framework/Platform/SDL2Window_Windowing.cs b/osu.Framework/Platform/SDL2Window_Windowing.cs
index debf407df..1adb46472 100644
--- a/osu.Framework/Platform/SDL2Window_Windowing.cs
+++ b/osu.Framework/Platform/SDL2Window_Windowing.cs
@@ -27,6 +27,7 @@ private void setupWindowing(FrameworkConfigManager config)
             CurrentDisplayBindable.Default = PrimaryDisplay;
             CurrentDisplayBindable.ValueChanged += evt =>
             {
+                graphicsSurface.UpdateActiveDisplay(evt.NewValue);
                 windowDisplayIndexBindable.Value = (DisplayIndex)evt.NewValue.Index;
             };
 
```

cc @peppy

Comment (smoogipoo) [reply src/Veldrid/MTL/MTLGraphicsDevice.cs:65]: Is there a reason for putting it inside `SwapBuffers`? I'd probably put it in `BeginFrame`. That also goes for `AllowTearing` which is currently sitting arbitrarily inside `GameHost.DrawFrame()`

Comment (frenzibyte) [reply src/Veldrid/MTL/MTLGraphicsDevice.cs:65]: That's what I had at first, but unfortunately `WaitForNextFrameReady` is called before `BeginFrame`, so we'll have to move the call to happen inside `BeginFrame`, which doesn't sound optimal.

Comment (smoogipoo) [reply src/Veldrid/MTL/MTLGraphicsDevice.cs:65]: Oh... Probably want a separate method into `Renderer` from the exact point `WaitForNextFrameReady` is (i.e. it _must_ be before querying/spinning on the triple buffer). I still don't feel like `SwapBuffer` is the correct place.

Comment (peppy) [reply src/Veldrid/MTL/MTLGraphicsDevice.cs]: I don't understand @frenzibyte's concerns with the original logic.

That said, I think this whole approach is fundamentally flawed.

Logically: if a frame hasn't been *started*, the callback should not wait for it to *end*. So I've fixed this by making the semaphore manual reset, and added a reset call in `WaitForNextFrameReady`. I've tested this to work with the original proposed framework diff, and I've also cross-tested with https://github.com/ppy/osu-framework/pull/5846

**Review by peppy — APPROVED** [review]

File: src/Veldrid/MTL/MTLGraphicsDevice.cs (hunk 9 above)
Comment (frenzibyte) [reply src/Veldrid/MTL/MTLGraphicsDevice.cs]: > I don't understand @frenzibyte's concerns with the original logic.

Right, I checked back on what I said and it turns out I misunderstood the logic a bit. With the code above, the callback can only exit early if `UpdateActiveDisplay` was called in the middle of a draw frame, so it won't have any major effect on synchronisation. The case I was thinking of was if the callback waits longer than the next V-sync, at which desynchronisation happens I guess ([read attached image](https://discord.com/channels/188630481301012481/589331078574112768/1119684304096743424)).
